### PR TITLE
Feature/cdr 377

### PR DESCRIPTION
**Pull request by micha1970**

### CODE REVIEW

File: example-generator/src/main/java/org/ehrbase/openehr/sdk/examplegenerator/ExampleGeneratorConfig.java [hunk 1]
```diff
@@ -200,15 +204,19 @@ private static int determineIndex(double t, int size) {
...
     }
 
-    private static Optional<WebTemplateInput> getInput(WebTemplateNode node, String suffix) {
-        return Optional.ofNullable(node)
-                .map(WebTemplateNode::getInputs)
-                .stream()
-                .flatMap(List::stream)
-                .filter(e -> Objects.equals(suffix, e.getSuffix()))
-                .findFirst();
+    private static Optional<WebTemplateInput> getInput(WebTemplateNode node, Predicate<WebTemplateInput> predicate) {
```
Comment (HolgerReiseVSys) [inline example-generator/src/main/java/org/ehrbase/openehr/sdk/examplegenerator/ExampleGeneratorConfig.java:207]: Why dillute the signature?
The predicate seems to be always the same.

File: example-generator/src/main/java/org/ehrbase/openehr/sdk/examplegenerator/ExampleGeneratorConfig.java [hunk 2]
```diff
@@ -200,15 +204,19 @@ private static int determineIndex(double t, int size) {
...
+        return getInputs(node, predicate).stream().findFirst();
     }
-
+    
+    private static List<WebTemplateInput> getInputs(WebTemplateNode node, Predicate<WebTemplateInput> predicate) {
```
Comment (HolgerReiseVSys) [inline example-generator/src/main/java/org/ehrbase/openehr/sdk/examplegenerator/ExampleGeneratorConfig.java:211]: Do you really need that method.
It is only used once and the predicate is not used.
Effectively it spares you from calling once
```
Optional.ofNullable(node)
  .map(WebTemplateNode::getInputs)
  .stream()
```

File: example-generator/src/main/java/org/ehrbase/openehr/sdk/examplegenerator/ExampleGeneratorConfig.java [hunk 3]
```diff
@@ -312,13 +320,30 @@ static void handleIsmTransition(IsmTransition value, WebTemplateNode node) {
             Transition transition = transitions.iterator().next();
             State currentState = transition.getTargetState();
 
+            List<WebTemplateInput> cfs = getInputs(node.findChildById("careflow_step").orElseThrow(), t -> true);
```
Comment (HolgerReiseVSys) [inline example-generator/src/main/java/org/ehrbase/openehr/sdk/examplegenerator/ExampleGeneratorConfig.java:323]: Do we really expect multiple `careflow_step`s?

File: example-generator/src/main/java/org/ehrbase/openehr/sdk/examplegenerator/ExampleGeneratorConfig.java (hunk 1 above)
Comment (micha1970) [reply example-generator/src/main/java/org/ehrbase/openehr/sdk/examplegenerator/ExampleGeneratorConfig.java:207]: The getInputs is the more generic method now

File: example-generator/src/main/java/org/ehrbase/openehr/sdk/examplegenerator/ExampleGeneratorConfig.java (hunk 2 above)
Comment (micha1970) [reply example-generator/src/main/java/org/ehrbase/openehr/sdk/examplegenerator/ExampleGeneratorConfig.java:211]: For me is was more clear to have this method than just inlining

File: example-generator/src/main/java/org/ehrbase/openehr/sdk/examplegenerator/ExampleGeneratorConfig.java (hunk 3 above)
Comment (micha1970) [reply example-generator/src/main/java/org/ehrbase/openehr/sdk/examplegenerator/ExampleGeneratorConfig.java:323]: As fae as I know, a web template can configure multiple valid careflow_steps!

Comment (HolgerReiseVSys) [reply example-generator/src/main/java/org/ehrbase/openehr/sdk/examplegenerator/ExampleGeneratorConfig.java]: There is of cause one careflow_step WebTemplateNode in the webtemplate with 0..1 instances in the IsmTransition.
There should be only one input with name "code" in the "careflow_step" node, as with "transition".

It careflow_steps are available, they must constrain the available states before selecting current_state and transition.
Also, the selected careflow_step needs to be constrained by the available states.
Thisway current_state, transition and careflow_step all agree on the state.

File: example-generator/src/main/java/org/ehrbase/openehr/sdk/examplegenerator/ExampleGeneratorConfig.java [hunk 4]
```diff
@@ -268,52 +272,143 @@ static void handleInstructionDetails(InstructionDetails value, WebTemplateNode n
...
+            .stream()
+            .flatMap(l -> l.stream())
+            .collect(Collectors.toSet());
+          
+          //1. check if there is a match between careflow_step and transition related to a current_state
+          Optional<Triple<Pair<String, WebTemplateInputValue>,State,Transition>> candidate = allCurrentStates.stream()
+            .map(cs -> {
+              String code = cs.getCode();
+              Set<Pair<String,WebTemplateInputValue>> filteredCfs = filterCareflowSteps(allCareflowTerminologyPairs, code);
+              Set<Transition> filteredTrs = filterTransitions(allTransitions, code);
```
Comment (HolgerReiseVSys) [inline example-generator/src/main/java/org/ehrbase/openehr/sdk/examplegenerator/ExampleGeneratorConfig.java:362]: `::filterTransitions` is only halled here. Why the detour State -> String -> State?

File: example-generator/src/main/java/org/ehrbase/openehr/sdk/examplegenerator/ExampleGeneratorConfig.java [hunk 5]
```diff
@@ -268,52 +272,143 @@ static void handleInstructionDetails(InstructionDetails value, WebTemplateNode n
...
+              //we need just one
+              if(filteredCfs.isEmpty() || filteredTrs.isEmpty())
+                return null;
+              return Triple.of(filteredCfs.iterator().next(), cs, filteredTrs.iterator().next());
+            })
+            .filter(t -> t != null)
```
Comment (HolgerReiseVSys) [inline example-generator/src/main/java/org/ehrbase/openehr/sdk/examplegenerator/ExampleGeneratorConfig.java:368]: Could use `Objects::nonNull`

File: example-generator/src/main/java/org/ehrbase/openehr/sdk/examplegenerator/ExampleGeneratorConfig.java [hunk 6]
```diff
@@ -268,52 +272,143 @@ static void handleInstructionDetails(InstructionDetails value, WebTemplateNode n
...
+            .collect(Collectors.toSet());
+          
+          //1. check if there is a match between careflow_step and transition related to a current_state
+          Optional<Triple<Pair<String, WebTemplateInputValue>,State,Transition>> candidate = allCurrentStates.stream()
+            .map(cs -> {
+              String code = cs.getCode();
+              Set<Pair<String,WebTemplateInputValue>> filteredCfs = filterCareflowSteps(allCareflowTerminologyPairs, code);
+              Set<Transition> filteredTrs = filterTransitions(allTransitions, code);
+              //we need just one
+              if(filteredCfs.isEmpty() || filteredTrs.isEmpty())
```
Comment (HolgerReiseVSys) [inline example-generator/src/main/java/org/ehrbase/openehr/sdk/examplegenerator/ExampleGeneratorConfig.java:364]: The candidate approach is great.
A simplification could be to include these values in the candidates 
and sort by left != null and then right != null.
Then fill the ismTransition from the first candidate.

**Review by HolgerReiseVSys — CHANGES_REQUESTED** [review]

File: example-generator/src/main/java/org/ehrbase/openehr/sdk/examplegenerator/ExampleGeneratorConfig.java [hunk 7]
```diff
@@ -268,52 +272,143 @@ static void handleInstructionDetails(InstructionDetails value, WebTemplateNode n
...
-                transitions.removeIf(t -> !states.contains(t.getTargetState()));
-            }
-
-            //take first transition
-            Transition transition = transitions.iterator().next();
-            State currentState = transition.getTargetState();
-
-            value.setCurrentState(currentState.toCodedText());
-            value.setTransition(transition.toCodedText());
+          Set<WebTemplateInputValue> theCurrentStates = getInput(node.findChildById("current_state").orElseThrow(), "code")
```
[the remark this answers is not in the document]
Comment (micha1970) [reply example-generator/src/main/java/org/ehrbase/openehr/sdk/examplegenerator/ExampleGeneratorConfig.java:336]: Good idea to put it into one method. The empty set is already considered

File: example-generator/src/main/java/org/ehrbase/openehr/sdk/examplegenerator/ExampleGeneratorConfig.java (hunk 6 above)
Comment (micha1970) [reply example-generator/src/main/java/org/ehrbase/openehr/sdk/examplegenerator/ExampleGeneratorConfig.java:364]: Yes that's an option, but to be honest, I prefer it this way

File: example-generator/src/main/java/org/ehrbase/openehr/sdk/examplegenerator/ExampleGeneratorConfig.java [hunk 8]
```diff
@@ -268,52 +272,136 @@ static void handleInstructionDetails(InstructionDetails value, WebTemplateNode n
...
+            .filter(Objects::nonNull)
+            .findFirst();
+          
+          if(candidate.isPresent()) {
+            //we got a match use it and return immediately
+            candidate.ifPresent(c -> fillIsmTransitionByTriple(value, c));
+            return;
+          }
+          
+          //try careflow_step not empty, look for a creflow_step with a valid current_state
```
Comment (HolgerReiseVSys) [inline example-generator/src/main/java/org/ehrbase/openehr/sdk/examplegenerator/ExampleGeneratorConfig.java:370]: type `creflow_step`

**Review by HolgerReiseVSys — CHANGES_REQUESTED** [review]

**Review by HolgerReiseVSys — APPROVED** [review]

**Pull request merged**